modules/custom/falcon_commerce/src/Plugin/rest/resource/DonationResource.php

```diff
@@ -0,0 +1,294 @@
+    $this->logger->info($request->getContent());
+
+    // TODO: check if we can get this data some other way provided by REST.
+    // Look into the serialization_class.
+    $this->data = json_decode($request->getContent(), TRUE);
+
+    $order = $this->createOrder();
+
+    // TODO: add payment processing.
+    // $this->>processPayment($order);
+
+    $response = new ResourceResponse($order, 200);
+
+    return $response;
+  }
+
+  /**
+   * Creates an order from given data.
+   */
+  protected function createOrder() {
+    $store_ids = [];
+    $order_items = [];
+    $variation_sku = 'donation';
+    /** @var \Drupal\commerce_product\Entity\ProductVariationInterface $variation */
+    $variations = $this->entityTypeManager->getStorage('commerce_product_variation')->loadByProperties(['sku' => $variation_sku]);
+    $variation = reset($variations);
+
+    if (!$store_ids) {
+      $product = $variation->getProduct();
+      $store_ids = $product->getStoreIds();
+    }
+
+    // Create order item object.
+    /** @var \Drupal\commerce_order\Entity\OrderItemInterface $order_item */
+    $order_item = $this->entityTypeManager->getStorage('commerce_order_item')->create([
+      'title' => $this->t('Donation'),
+      'type' => 'donation',
+      'purchased_entity' => $variation,
+      'quantity' => 1,
+      'field_donation_type' => $this->data['donation_type'],
+      'appeal' => $this->data['appeal'],
+    ]);
+
+    // Set price.
+    $unit_price = new Price($this->data['payment']['amount'], $this->data['payment']['currency_code']);
+    $order_item->setUnitPrice($unit_price, TRUE);
+
+    try {
+      // We need only entity level violations.
+      /** @var \Drupal\Core\Entity\EntityConstraintViolationListInterface $violations */
+      $violations = $order_item->validate()->getEntityViolations();
+      if ($violations->count() > 0) {
+        foreach ($violations as $violation) {
+          $this->logger->info($violation->getPropertyPath() ?: 'Entity' . ': ' . $violation->getMessage());
+        }
+
+        throw new \Exception('Order item violations.');
+      }
+      $order_item->save();
+    }
+    catch (\Exception $e) {
+      watchdog_exception('falcon_commerce', $e);
+      throw new HttpException(500, 'Internal Server Error', $e);
+    }
+
+    $order_items[] = $order_item;
+
+    // Get user and profile objects.
+    $account = $this->getUser();
+    $profile = $this->getProfile();
+
+    // Create an order.
+    /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+    $order = $this->entityTypeManager->getStorage('commerce_order')->create([
+      'type' => 'donation',
+      'state' => 'draft',
+      'mail' => $account->getEmail(),
+      'uid' => $account->id(),
+      'billing_profile' => $profile,
+      'store_id' => reset($store_ids),
+      'order_items' => $order_items,
+      'placed' => $this->time->getRequestTime(),
+      'field_appeal' => $this->data['appeal'],
+    ]);
```

Comment:
How would we same any extra fields? (i.e. source code from ThankQ module)

Reply:
Wrapped order related fields into "order" key.